File: srcs/requirements/Pong-Server/server/game/AI.py
from config import *
from Vec2 import *
from Hitbox import *
import math

class AI:
	def __init__(self, player):
		self.id = player.nb
		self.side = player.side
		self.size = player.size
		self.time = 0
		self.pos = Vec2(pos=player.paddle[0].pos)
		self.center = Vec2(pos=self.pos)
		self.target = Vec2(pos=self.pos)
		self.moves = []
		if self.side == "left" or self.side == "right":
			self.hitbox = Hitbox(pos=Vec2(self.pos.x - 1, 0), width=player.size[0] + 2, height=winHeight)
		else:
			self.hitbox = Hitbox(pos=Vec2(0, self.pos.y - 1), width=winWidth, height=player.size[1] + 2)

	def update(self, core, core_delta):
		if core.ball.stick == self.id:
			self.moves.append("LAUNCH")
		else:
			tmp = time.time()
			delta = tmp - self.time
			if delta >= 1:
				self.predict(core, core_delta)
				self.time = tmp
			
			if self.target != self.pos:
				self.move()
		if self.target != self.pos:
			if self.closeEnough():
				self.target = Vec2(self.pos.x, self.pos.y)

	def predict(self, core, delta):
		sec_to_predict = 5

		ball_cpy = core.ball.copy(self.hitbox)

		collided = False
		for i in range(sec_to_predict * 100):
			ball_cpy.update(core, delta)
			ball_box = [ball_cpy.center[0].x - ball_cpy.radius, ball_cpy.center[0].y - ball_cpy.radius]
			if is_colliding(ball_box, [ball_cpy.radius * 2, ball_cpy.radius * 2], self.hitbox.pos, self.hitbox.size):
				collided = True
				if is_colliding(ball_box, [ball_cpy.radius * 2, ball_cpy.radius * 2], self.pos, self.size):
					break
				if self.side == "left" or self.side == "right":
					self.target = Vec2(self.pos.x, ball_cpy.center[0].y - (self.size[1] / 2))
				else:
					self.target = Vec2(ball_cpy.center[0].x - (self.size[0] / 2), self.pos.y)
				break

		if not collided:
			self.target = Vec2(pos=self.center)

		if is_colliding(ball_box, [ball_cpy.radius * 2, ball_cpy.radius * 2], self.pos, self.size):
			self.target = Vec2(pos=self.pos)
		elif self.target != self.center or (self.target == self.center and collided):
			if self.side == "left" or self.side == "right":
				self.target.y += randint(-(self.size[1] // 2 - int(self.size[1] * 0.07)), (self.size[1] // 2 - int(self.size[1] * 0.07)))
			else:
				self.target.x += randint(-(self.size[0] // 2 - int(self.size[0] * 0.07)), (self.size[0] // 2 - int(self.size[0] * 0.07)))
  

	def move(self):
		if self.target.x < self.pos.x:
			self.moves.append("LEFT")
		elif self.target.x > self.pos.x:
			self.moves.append("RIGHT")
		elif self.target.y < self.pos.y:
			self.moves.append("UP")
		elif self.target.y > self.pos.y:	
			self.moves.append("DOWN")
   
	def closeEnough(self):
		if self.side == "left" or self.side == "right":
			if abs(self.target.y - self.pos.y) < math.ceil(self.size[1] * 0.07):
				return True
		else:
			if abs(self.target.x - self.pos.x) < math.ceil(self.size[0] * 0.07):
				return True
		return False

File: srcs/django_code/front/js/game/AI.js
import { Vec2 } from './Vec2.js'
import { Hitbox, is_colliding } from './Hitbox.js'
import { canvas } from './canvas.js';

export class AI {
	constructor(player) {
		this.id = player.nb;
		this.side = player.side;
		this.size = player.size;
		this.time = 0;
		this.pos = new Vec2(player.paddle[0].pos.x, player.paddle[0].pos.y);
		this.center = new Vec2(this.pos.x, this.pos.y);
		this.target = new Vec2(this.pos.x, this.pos.y);
		this.moves = [];
		if (this.side === "left" || this.side === "right")
			this.hitbox = new Hitbox(this.pos.x - 1, 0, player.size[0] + 2,  canvas.height)
		else
			this.hitbox = new Hitbox(0, this.pos.y - 1, canvas.width, player.size[1] + 2)
	}

	update(core, core_delta) {
		if (core.ball.stick === this.id)
			this.moves.push("LAUNCH");
		else {
			const tmp = Date.now() / 1000;
			const delta = tmp - this.time;
			if (delta >= 1) {
				this.predict(core, core_delta);
				this.time = tmp;
			}
			if (this.target.notEqual(this.pos))
				this.move();
		}
		if (this.target.notEqual(this.pos))
			if (this.closeEnough())
				this.target = new Vec2(this.pos.x, this.pos.y);
	}

	predict(core, delta) {
		const sec_to_predict = 5;

		let ball_cpy = core.ball.copy(this.hitbox);
		let ball_box = new Vec2(ball_cpy.center[0].x - ball_cpy.radius, ball_cpy.center[0].y - ball_cpy.radius);
		let collided = false;
		for (let i = 0; i < sec_to_predict * 100; i++) {
			ball_cpy.update(core, delta);
			ball_box = new Vec2(ball_cpy.center[0].x - ball_cpy.radius, ball_cpy.center[0].y - ball_cpy.radius);
			if (is_colliding(ball_box, [ball_cpy.radius * 2, ball_cpy.radius * 2], this.hitbox.pos, this.hitbox.size)) {
				collided = true;
				if (is_colliding(ball_box, [ball_cpy.radius * 2, ball_cpy.radius * 2], this.pos, this.size))
					break;
				if (this.side === "left" || this.side === "right")
					this.target.y = ball_cpy.center[0].y - (this.size[1] / 2);
				else
					this.target.x = ball_cpy.center[0].x - (this.size[0] / 2);
				break;
			}
		}
		if (!collided)
			this.target = new Vec2(this.center.x, this.center.y);

		if (is_colliding(ball_box, [ball_cpy.radius * 2, ball_cpy.radius * 2], this.pos, this.size))
			this.target = new Vec2(this.pos.x, this.pos.y);
		else if (this.target.notEqual(this.center) || (this.target.equal(this.center) && collided)) {
			if (this.side === "left" || this.side === "right")
				this.target.y += randHit(this.size[1]);
			else
				this.target.x += randHit(this.size[0]);
		}
	}

	move() {
		if (this.target.x < this.pos.x)
			this.moves.push("LEFT");
		else if (this.target.x > this.pos.x)
			this.moves.push("RIGHT");
		else if (this.target.y < this.pos.y)
			this.moves.push("UP");
		else if (this.target.y > this.pos.y)
			this.moves.push("DOWN");
	}

	responsive(players, old_sizes) {
		for (let player of players) {
			if (this.id == player.nb) {
				this.size = player.size;
				this.pos = new Vec2(player.paddle[0].pos.x, player.paddle[0].pos.y);
				break;
			}
		}
		let pos_ratio = [this.center.x / old_sizes[0], this.center.y / old_sizes[1]];
		this.center = new Vec2(pos_ratio[0] * canvas.width, pos_ratio[1] * canvas.height);
		pos_ratio = [this.target.x / old_sizes[0], this.target.y / old_sizes[1]];
		this.target = new Vec2(pos_ratio[0] * canvas.width, pos_ratio[1] * canvas.height);
		if (this.side === "left" || this.side === "right")
			this.hitbox = new Hitbox(this.pos.x - 1, 0, this.size[0] + 2,  canvas.height)
		else
			this.hitbox = new Hitbox(0, this.pos.y - 1, canvas.width, this.size[1] + 2)
	}

	closeEnough() {
		if (this.side === "left" || this.side === "right") {
			if (Math.abs(this.target.y - this.pos.y) < (this.size[1] * 0.05))
				return true;
		}
		else {
			if (Math.abs(this.target.x - this.pos.x) < (this.size[0] * 0.05))
				return true;
		}
		return false;
	}
}

function randHit(size) {
	const max = size / 2 - (size * 0.07);
	const min = -max;
	return Math.floor(Math.random() * (max - min + 1) + min);
}